### PR TITLE
Connection: disconnect connection owner upon removal

### PR DESCRIPTION
**Pull request by sergeymitr**
When connected users are removed API requests are sent to WPCOM to unlink them.
It's already hooked into `deleted_user` and `remove_user_from_blog`, and works fine.

However, there are two limitations:
1. It doesn't happen for connection owners, they are skipped. Secondary users are disconnected just fine.
2. The hooks are initialized in Jetpack plugin, so standalone plugins are ignored.

## Proposed changes:
* Make sure connection owner is disconnected on WPCOM if their WordPress user account is removed.
* Move the hooks from Jetpack plugin into Connection package to also cover standalone plugins.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
p1712570085579179-slack-CDLH4C1UZ

## Does this pull request change what data or activity we track or use?
No.

## Testing instructions:
1. Connect Jetpack under one WordPress user account (user A).
2. Create another account (user B), logout and login under that user.
3. Go to "Users" and delete the user A (connection owner). Ignore the "Connect to WordPress.com" banner.
4. Run Jetpack Debugger over the site. Confirm that it says `Everything looks great! (site-only connection)`

### CODE REVIEW

File: projects/packages/connection/src/class-manager.php [hunk 1]
```diff
@@ -863,6 +867,22 @@ public function connect_user( $user_id = null, $redirect_url = null ) {
...
 	}
 
+	/**
+	 * Force user disconnect.
+	 *
+	 * @param int $user_id Local (external) user ID.
+	 *
+	 * @return bool
+	 */
+	public function disconnect_user_force( $user_id ) {
```
Comment (fgiannar) [inline projects/packages/connection/src/class-manager.php:884]: Nitpick but `force_disconnect_user` sounds a bit more intuitive :)

**Review by fgiannar — CHANGES_REQUESTED** [review]
Thanks for working on this, Sergey!

One potential issue with the current implementation, is that when we have an older version of the Jetpack plugin and a newer version of a standalone, eg Backups, `disconnect_user` will be called twice and the remote call to `jetpack.unlink_user` will occur twice as well.
Maybe we could update the hook priorities to `9` when registering the corresponding `deleted_user` and `remove_user_from_blog` actions and add an extra check in `disconnect_user` to return early if `did_action('jetpack_unlinked_user')` and/or the corresponding user is not connected?

File: projects/packages/connection/src/class-manager.php (hunk 1 above)
Comment (sergeymitr) [reply projects/packages/connection/src/class-manager.php:884]: I believe `disconnect_user_force` is more in line with the naming we already follow in the `Manager` class: `disconnect_site()` and `disconnect_site_wpcom()`, `get_connected_*`, `handle_registration()` and `handle_authorization()`, etc.
We always put the action first, and `_force` would be a descriptor that clarifies how exactly we disconnect the user here.

**Comment by sergeymitr** [comment]
Hi @fgiannar, thanks for taking a look!
That's a good catch.

Unfortunately `did_action()` doesn't allow us to check which user the action was called for, so it would cause issues with bulk delete of users, and introduce some room for future hard-to-track issues.

Besides, the issue will only affect connected secondary users on sites with a latest standalone plugin with an outdated Jetpack, and deleting a user is not such a common action either way.
But most importantly, calling the method (and the endpoint) twice doesn't cause any issues, so I think there's no need to add more complexity for such a rare edge case.

**Comment by sergeymitr** [comment]
> or the corresponding user is not connected

That will have its own side effects, such as leaving users connected on WPCOM if they are missing user token.
So I'd rather avoid that as well.

File: projects/packages/connection/src/class-manager.php (hunk 1 above)
Comment (fgiannar) [reply projects/packages/connection/src/class-manager.php:884]: Yeah in this case the action would be `force_disconnect` (verb), but anyway it's definitely not a blocker.

**Comment by fgiannar** [comment]
> > or the corresponding user is not connected
> 
> That will have its own side effects, such as leaving users connected on WPCOM if they are missing user token. So I'd rather avoid that as well.

WPCOM has self-healing mechanisms that will remove the token if it doesn't exist at the remote site. On top of that, I think having old versions of the Jetpack plugin co-existing with newer versions of other standalones is not an edge case anymore. For all these cases, we would be triggering a redundant remote request. 
I think that it makes more sense to not attempt to disconnect an already disconnected user as in normal flows the user will be disconnected on WPCOM the first time we disconnect them at the remote site.

**Comment by sergeymitr** [comment]
@fgiannar that's a good point.
I added the additional check, please give it another look.

**Review by fgiannar — APPROVED** [review]
Excellent 👍 Many thanks for working on this! 🚢

**Pull request merged**